Minimize rootfs partition size, less flash activity

Doridian opened the pull request:
This proposes the following changes:
- Change rootfs to ext2 (for a readonly FS we have no need for all the features of ext3/4 like journaling), this saves some space for journal inodes etc
- Use resize2fs and other utilities to reduce the size of `rootfs.ext2` to the smallest possible, this means flashes go quicker as it only has to re-write the portion of the rootfs partition that actually contains data (allows you to reserve a large space for rootfs for future expansion, but not use it immediately and slow down updates unnecessarily)

bboehmke commented:
Based on my knowledge ext4 has some features that gives us better performance.
Also I think ext2 only supports fixed block size which could cause more disk usage as needed.
So I think we should stay with ext4 and maybe disable the journaling because thats for sure not required/used for a read only system.

The second part (reduce image size to content) sounds fine. A faster update process is always great.

@a16bitsysop What do you think?

Doridian commented:
@bboehmke Thanks for the input with the features. I have updated my branch/PR to generate an ext4 FS instead, but without the features we don't need.
According to the manual for genimage at https://github.com/pengutronix/genimage#ext2-ext3-ext4
The default options for `ext4` are `extents,uninit_bg,dir_index,has_journal`.
We have no use for `has_journal` or `uninit_bg`, as both of these are only useful on r/w filesystems.
So I changed it to only enable `extents, dir_index`

a16bitsysop commented:
Yes I think ext4 is better, I couldn’t see this as a PR only the README one.

But also I was looking at moving things around to make a ‘standard’ image as well as per issue #4, so with that option it would build an image with one rootfs that is read write and no datafs.  Unless it’s a really big saving it’s better to have as much recovery options as possible, would also need no journal options so can be overridden for ‘standard’ image.

It’s good to have a bit of a buffer for image size as a simple “mount -o remount,rw /“ makes it possible to add an extra package or something without having to rebuild the image if you are still testing and setting up.  I saw some read only images have a “rw” and “ro” script to make the rootfs writable or read only.

> On 7 Dec 2021, at 11:48, Benjamin Böhmke ***@***.***> wrote:
> 
> ﻿
> Based on my knowledge ext4 has some features that gives us better performance.
> Also I think ext2 only supports fixed block size which could cause more disk usage as needed.
> So I think we should stay with ext4 and maybe disable the journaling because thats for sure not required/used for a read only system.
> 
> The second part (reduce image size to content) sounds fine. A faster update process is always great.
> 
> @a16bitsysop What do you think?
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or unsubscribe.
> Triage notifications on the go with GitHub Mobile for iOS or Android.

a16bitsysop commented:
I can see it now, I was looking on gitlab

bboehmke commented:
Yes that is a good point, for development purpose it would be great if the file system has some free space.
Maybe we can make it configurable.

For example if the `SIZE_ROOT_FS` is 0 we use the `SIZE_ROOT_PART` as size for the file system and shrink it after creation. If a value for `SIZE_ROOT_FS` is given we keep it as is without shrinking.

a16bitsysop commented:
Yes it would be good to shrink if `SIZE_ROOT_FS` is 0.

Apart from that it looks okay and can be reworked to create a "standard" image as well in the future.

Doridian commented:
Added an `if` so that if `SIZE_ROOT_FS` is 0, we shrink it, otherwise we go back to the old code path entirely

a16bitsysop commented:
Thats good, is it actual 0 free bytes after shrink? I think it would be good to allow 8mb or a nominal amount after shrinking so configuration can be changed in an emergency without burning a new image.

Also change the armv7 test so `--env SIZE_ROOT_FS="150M"` is 0, then it gets tested by the CI as well.

Doridian commented:
> Thats good, is it actual 0 free bytes after shrink? I think it would be good to allow 8mb or a nominal amount after shrinking so configuration can be changed in an emergency without burning a new image.
> 
> Also change the armv7 test so `--env SIZE_ROOT_FS="150M"` is 0, then it gets tested by the CI as well.

The actual free space is indeed 0. You might be able to write a tiny file, but unlikely.
Leaving X free space would require some more complex calculations (as in, actually calculating minimum size, then adding 8MB), which I am not comfortable enough with bash scripting to do that in any reasonable way.
If anyone feels up to the task, feel free to extend my PR for covering such.

On the note of the tests: I have adjusted the test as requested.

bboehmke commented:
Yeah I would say to leave it like this for now and extend it late if needed.
For me it look good to merge now.

a16bitsysop commented:
Yes it's fine like this, it has both options available as well.

a16bitsysop commented:
I had a look and think it easier just to create and empty file of nominal size, I couldn't delete it from the image after shrinking during the build process

https://gitlab.com/raspi-alpine/builder/-/merge_requests/60
